Initial Commit For ReSharper Migration Guide (#154)
JetBrains/jetbrains_guide: sites/dotnet-guide/contents/tutorials/resharper-to-rider/nuget-package-management/index.md

---
type: TutorialStep

> date: 2021-01-29

Reviewer: ```suggestion
date: 2021-03-29
```

title: NuGet Package Management
technologies: [.net,csharp]
products: [rider,resharper]
topics: [ide]
author: khalidabuhakmeh
subtitle: Learn the basics of the Rider's NuGet Tool Window from a Visual Studio + ReSharper user's perspective.
thumbnail: ./thumbnail.png
---

> We can find the greatest strength of every modern development stack in its community, more precisely, in its open-source offerings. For .NET Developers, we can access the breadth of community know-how through the NuGet package service. Visual Studio users will likely be familiar with the NuGet tool window, where developers can manage existing dependencies or add new ones.

Reviewer: ```suggestion
We can find the greatest strength of every modern development stack in its community. More precisely, in its open-source offerings. For .NET Developers, we can access the breadth of community know-how through NuGet. Visual Studio users will likely be familiar with the NuGet tool window, where developers can manage existing dependencies or add new ones.
```

![Visual Studio's NuGet Tool Window](./1-visual-studio-nuget-window.png)

In Visual Studio's NuGet tool window, we can see the package's name, its authors, current version, license, and additional metadata to help the user understand more about the dependency. We can also manage the sources Visual Studio will use to search and download the packages.

> When we install a package, the NuGet process will update the appropriate project (.csproj) targets with our selected package's information.

Reviewer: ```suggestion
When we install a package, the NuGet process will update the appropriate project (`.csproj`) targets with our selected package's information.
```

![JetBrains Rider's NuGet Tool Window](./2-rider-nuget-tool-window.png)

Rider offers several approaches to help developers manage dependencies. Let's start with the most common method, using the **NuGet Tool Window**. We will find the **NuGet Tool Window** tab at the bottom of our IDE instance. Clicking the tab will show a UI that includes a search bar, results, installed packages, and additional metadata.

![JetBrains Rider's NuGet Tool Context Menu With Actions](./3-rider-nuget-tool-window-context-menu.png)

We can add a package to the entire solution or selectively choose the projects that require a specific package from the tool window. If we're still unsure about a particular package, we can use the <kbd>Alt+Enter</kbd> shortcut to toggle a context menu that will display several actions used to help us learn more about a package:

- Quick NuGet Documentation
- Open In Assembly Explorer
- Show Package on NuGet.Org
- Show Project Package
- Show License Page
- Search Package with Google

> Our personal favorite is to use the **Open In AssemblyExplorer** action, downloading the package and loading it in Rider's world-class **AssemblyExplorer**. We can explore the contents of any package and be sure it matches our expectations.

Reviewer: ```suggestion
Our personal favorite is to use the **Open In Assembly Explorer** action, downloading the package and loading it in Rider's world-class **Assembly Explorer**. We can explore the contents of any package and be sure it matches our expectations.
```

![JetBrains Rider NuGet Tool Sources and Feeds](./6-rider-nuget-feeds.png)

As we've seen, the Rider NuGet Tool Window is more than capable of managing existing NuGet dependencies and finding new OSS packages to help solve any problem. But there is still one more thing that differentiates Rider's NuGet tool from the one found in Visual Studio. That feature is **Type search**.  Using the `T:` prefix in our NuGet search bar will find all packages that contain a Type definition.

![JetBrains Rider NuGet Type lookup](./7-rider-type-lookup.png)

This feature is also available in the editor. We can recall types from memory and then use a quick action to **Find the type on NuGet.org...*, allowing us to stay in the flow and continue writing code.

![JetBrains Rider NuGet Type lookup in editor](./8-rider-type-lookup-editor.png)

When working with NuGet packages, Rider gives developers options, regardless of your workflow. Whether we want to perform some exhaustive research or know precisely what package we need, Rider is there to help.

### See Also

> - [NuGet Tool Window in Rider](/tutorials/rider-essentials/nuget/)

Reviewer: Can you make this the full URL (and double check all links in the PR)?

Production is under `jetbrains.com/dotnet/guide`, so linking to `/` will 404 on deployment.
